packages/flutter_tools/lib/src/commands/clean.dart

```diff
@@ -81,11 +88,22 @@ class CleanCommand extends FlutterCommand {
     try {
       final XcodeProjectInterpreter xcodeProjectInterpreter = globals.xcodeProjectInterpreter!;
       final XcodeProjectInfo projectInfo = (await xcodeProjectInterpreter.getInfo(xcodeWorkspace.parent.path))!;
-      for (final String scheme in projectInfo.schemes) {
+      if (argResults?.wasParsed('scheme') ?? false) {
+        final String scheme = argResults!['scheme'] as String;
+        if (scheme.isEmpty) {
+          throwToolExit('Scheme given in option is empty !');
+        }
+        if (!projectInfo.schemes.contains(scheme)) {
+          throwToolExit('Scheme "$scheme" not found in ${projectInfo.schemes}');
+        }
         await xcodeProjectInterpreter.cleanWorkspace(xcodeWorkspace.path, scheme, verbose: _verbose);
+      } else {
+        for (final String scheme in projectInfo.schemes) {
+          await xcodeProjectInterpreter.cleanWorkspace(xcodeWorkspace.path, scheme, verbose: _verbose);
+        }
       }
     } on Exception catch (error) {
```

Comment:
Exception here was hidden.  If xcode didn't find scheme or throw something it's important to abort no ?  I've edited that.

Reply:
@christopherfujino Do you have any opinions on this? In particular, is it appropriate to throw a ToolExit in the case of input error (i.e. no scheme given)?

Reply:
> @christopherfujino Do you have any opinions on this? In particular, is it appropriate to throw a ToolExit in the case of input error (i.e. no scheme given)?

I'm not sure I understand the question. Are you asking if we should printError and exit 0 when the user provides a non-existent scheme? If so, then I think tool exiting is the right course of action (ToolExit generally means that we cannot continue because of some invalid state, usually based on user inputs).

If I'm misunderstanding the question, please clarify.

Reply:
Three new `ToolExit` throws are added here:

https://github.com/flutter/flutter/blob/8c7682a933a9fb856631706da2e9e1fb49bf1d18/packages/flutter_tools/lib/src/commands/clean.dart#L91-L96

and

https://github.com/flutter/flutter/blob/8c7682a933a9fb856631706da2e9e1fb49bf1d18/packages/flutter_tools/lib/src/commands/clean.dart#L103-L105

(side note: the first two shown in the first code snippet will get caught by the `catch` in the second code snippet, in which case a different `ToolExit` is thrown)

However, in other cases throughout the tool, `debug.printError` is used without throwing a `ToolExit`. Example:

https://github.com/flutter/flutter/blob/378387b139386a0c1e089c85a8e0bcd769dcfe47/packages/flutter_tools/lib/src/commands/clean.dart#L101-L104

Going back to the three newly added `ToolExit` throws, is it appropriate to throw here instead of just printing an error message? Asked more specifically:

1. Is it appropriate to throw if a user provides an invalid Xcode scheme name (or doesn't provide one at all despite using `--scheme`), or should we print an error message and continue with cleaning?
2. Is it appropriate to throw if something goes wrong with cleaning a workspace, or should we print an error message and continue with cleaning?

Reply:
> Three new `ToolExit` throws are added here:
> 
> https://github.com/flutter/flutter/blob/8c7682a933a9fb856631706da2e9e1fb49bf1d18/packages/flutter_tools/lib/src/commands/clean.dart#L91-L96
> 
> and
> 
> https://github.com/flutter/flutter/blob/8c7682a933a9fb856631706da2e9e1fb49bf1d18/packages/flutter_tools/lib/src/commands/clean.dart#L103-L105
> 
> (side note: the first two shown in the first code snippet will get caught by the `catch` in the second code snippet, in which case a different `ToolExit` is thrown)
> 
> However, in other cases throughout the tool, `debug.printError` is used without throwing a `ToolExit`. Example:
> 
> https://github.com/flutter/flutter/blob/378387b139386a0c1e089c85a8e0bcd769dcfe47/packages/flutter_tools/lib/src/commands/clean.dart#L101-L104
> 
> Going back to the three newly added `ToolExit` throws, is it appropriate to throw here instead of just printing an error message? Asked more specifically:
> 
> 1. Is it appropriate to throw if a user provides an invalid Xcode scheme name (or doesn't provide one at all despite using `--scheme`), or should we print an error message and continue with cleaning?

This to me seems like a clear user error, and thus should be a tool exit.

> 2. Is it appropriate to throw if something goes wrong with cleaning a workspace, or should we print an error message and continue with cleaning?

It makes sense to me printError and continue cleaning, but I don't have a strong opinion either way.